### ISSUE TITLE
Limit Concurrent Builds

### ISSUE THREAD

**Issue by lsenft** [Block8/PHPCI#517]
Is there a way to limit concurrent builds?  We are currently on a resource constrained server and have run into some issues when a dev pushes several commits at once.

Also is there a better place for community support?  Not sure this is ticket worthy, but I didn't see any forums, etc.

**Comment by SamMousa** [comment]
I think this is the default behavior now: https://github.com/Block8/PHPCI/blob/master/PHPCI/Command/RunCommand.php#L104

**Comment by elboletaire** [comment]
@SamMousa It seems, but sometimes it fires multiple times at a time.

Yesterday I was doing some tests using a local phpci and I tried to add multiple builds and tried to execute the `phpci:run-builds` multiple times at a time. Every time I've tried to run the `phpci:run-builds` it showed me a notice like "There's a currently running build".

Today I've updated phpci in my office server and after some commits there has been one push that triggered four builds at a time. A part from the fact that the server slowed down too much, some of that builds started showing this error:

```
fatal: destination path '/var/www/phpci/PHPCI/build/50' already exists and is not an empty directory.

Failed to clone remote git repository.
Exception: Could not create a working copy.
```

Probably is another issue, but started showing with this huge build.

In my case I'm using the normal cron for `phpci:run-builds` instead of the daemon.

**Comment by elboletaire** [comment]
As said in my previous comment:

![captura de 2015-04-07 17 59 03](https://cloud.githubusercontent.com/assets/153305/7027491/04a3b75a-dd50-11e4-8cce-5c40774de37e.png)
![captura de 2015-04-07 17 58 52](https://cloud.githubusercontent.com/assets/153305/7027490/04a2dad8-dd50-11e4-88b6-e9dd97bab4c7.png)

**Comment by elboletaire** [comment]
I think that I can confirm that when it starts making such a lot of builds, builds coming later crash with the following error(s):

```
fatal: destination path '/var/www/phpci/PHPCI/build/70' already exists and is not an empty directory.

Failed to clone remote git repository.
Exception: Could not create a working copy.
```

This is common too:

```
Cloning into '/var/www/phpci/PHPCI/build/78'...
error: cannot run /var/www/phpci/PHPCI/build/78.sh: El fitxer o directori no existeix
fatal: unable to fork

Exception: Warning: unlink(/var/www/phpci/PHPCI/build/78.key): No such file or directory in /var/www/phpci/PHPCI/Model/Build/RemoteGitBuild.php line 108
```
The sentence `El fitxer o directori no existeix` means "The file or folder does not exist".

After crashing this way, if I trigger a manual build everything seems to work as expected.

**Comment by Adirelle** [comment]
Do not start several run-builds at all. In the current state of the code, run-builds fetches a list of builds to run at the very start of its execution and try to build them all without refreshing its data between each build. Several run-builds in parallel will stomp on each other toes, trying to build the same jobs at nearly the same time.

**Comment by elboletaire** [comment]
And that's why it crashes?

What's the solution then? Create a custom script like described by @robertscherer?

**Comment by Adirelle** [comment]
That's it.

**Comment by Adirelle** [comment]
@elboletaire: also, if the `flock` command is available, you can try this (much simpler): 

```
flock --exclusive --nonblock /var/run/phpci.lock /var/www/phpci/console phpci:run-builds >> /var/log/phpci.log
```

This will exclusively lock the file /var/run/phpci.lock during the execution of phpci:run-builds, silently failing if the file is already locked (i.e. another instance of phpci:run-builds is running).

**Comment by elboletaire** [comment]
That's a little better.

**Comment by SamMousa** [comment]
Why not use a true task scheduler instead of trying to reinvent the wheel?
https://github.com/illuminate/queue 
For example would allow multiple runners relatively easy.

Also, there could be some pruning in each build, for example:
1. Build task starts.
2. Build task checks if the commit it must be run against fits the requirements (like "there must not be a commit on this branch, after the current commit, that has a build finished or in queue")
3. Build task runs

Since step 1 and 2 are fast it can skip task 3 if 2 is not ok; this will keep it simple and makes it so very little actual queue management is required.
